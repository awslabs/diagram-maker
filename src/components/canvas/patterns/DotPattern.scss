.dm-dots {
  shape-rendering: geometricPrecision;

  pattern {
    path {
      fill: #d5dbdb;
    }
  }
}
