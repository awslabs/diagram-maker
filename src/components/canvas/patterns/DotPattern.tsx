import * as Preact from 'preact';

export interface DotPatternProps {
  cellSize: number;
  radius: number;
}

const getQuarterCirclePathData = (cx: number, cy: number, r: number): string => {
  return `M${cx},${cy} m${-r},0 a${r},${r} 0 1,1 ${r * 2},0 a${r},${r} 0 1,1 ${-r * 2},0`;
};

const DotPattern = (props: DotPatternProps): JSX.Element => {
  const { cellSize, radius } = props;
  const pathTL = getQuarterCirclePathData(0, 0, radius);
  const pathTR = getQuarterCirclePathData(cellSize, 0, radius);
  const pathBR = getQuarterCirclePathData(cellSize, cellSize, radius);
  const pathBL = getQuarterCirclePathData(0, cellSize, radius);

  return (
    <svg className="dm-dots" xmlns="http://www.w3.org/2000/svg">
      <defs>
        <pattern id="dm-dot-pattern" width={cellSize} height={cellSize} patternUnits="userSpaceOnUse">
          <path d={pathTL} />
          <path d={pathTR} />
          <path d={pathBR} />
          <path d={pathBL} />
        </pattern>
      </defs>
      <rect fill="url(#dm-dot-pattern)" width="100%" height="100%" />
    </svg>
  );
};

export default DotPattern;
