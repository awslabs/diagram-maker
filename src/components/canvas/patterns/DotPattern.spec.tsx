import * as Preact from 'preact';
import { shallow } from 'preact-render-spy';

import DotPattern from './DotPattern';

describe('DotPattern', () => {
  it('renders dot pattern', () => {

    const dotPattern = shallow(
      <DotPattern cellSize={10} radius={1} />
    );

    expect(dotPattern).toMatchSnapshot();
  });
});
