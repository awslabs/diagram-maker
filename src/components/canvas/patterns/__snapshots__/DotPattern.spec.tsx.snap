// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DotPattern renders dot pattern 1`] = `
preact-render-spy (1 nodes)
-------
<svg
  class="dm-dots"
  xmlns="http://www.w3.org/2000/svg"
>
  <defs>
    <pattern
      id="dm-dot-pattern"
      width={10}
      height={10}
      patternUnits="userSpaceOnUse"
    >
      <path d="M0,0 m-1,0 a1,1 0 1,1 2,0 a1,1 0 1,1 -2,0"></path>
      <path d="M10,0 m-1,0 a1,1 0 1,1 2,0 a1,1 0 1,1 -2,0"></path>
      <path d="M10,10 m-1,0 a1,1 0 1,1 2,0 a1,1 0 1,1 -2,0"></path>
      <path d="M0,10 m-1,0 a1,1 0 1,1 2,0 a1,1 0 1,1 -2,0"></path>
    </pattern>
  </defs>
  <rect
    fill="url(#dm-dot-pattern)"
    width="100%"
    height="100%"
  >
  </rect>
</svg>

`;
